coastguard/clustersetnetworkpolicy.md

```diff
@@ -0,0 +1,272 @@
+As a clusterset implementation, we would still need to be aware of the possibility of localized single-cluster network policies.
+As such, we should not interfere with any such policies, should they be defined.
+Since these policies could inherently disallow inter-cluster traffic,
+our design and implementation should strive to complement them by adding support for such traffic.
+
+#### Pod IP Preservation
+
+For all practical applications of this solution, we would need to preserve the source IP in the packet 
+either by avoiding any NAT on the source cluster, or by utilizing a solution such as an overlay network.
+We would need to understand how this works with Globalnet v2 (if it's at all possible).
+
+### Design Tenets
+
+#### Broker Centric Data Sharing
+
+As we're running in the Submariner ecosystem, it makes sense to share any data we need to share via the broker.
+This has the following benefits (and more that aren't listed):
+
+* Debuggability - users can see and query the data on the broker with ease to trace any problems.
+* Standardized clients - we have standardized Kubernetes clients generated, with all the common features (selectors, etc).
+* Integrated RBAC and authentication - Kubernetes takes care of this aspect for us as well.
+
+#### Connectivity Decoupling
+
+We should strive to decouple the design and implementation from the underlying connectivity,
+building on top of the fact that connectivity exists but not counting on it to be implemented in some certain way.
+This would help our solution to be more robust and versatile.
+
+#### Solution Scope
+
+* We aim to solve the problem of having clusterset-aware network policies.
+* We're not trying to provide a revolutionary approach,
+  but rather an evolutionary approach that expands the base Kubernetes network policy concept.
+* We aim for a good enough solution, but not a perfect solution encompassing every possible use case and edge case.
+
+## Design Details
+
+### ClusterSetNetworkPolicy CRD
+
+A new `ClusterSetNetworkPolicy` CRD will be defined for specifying multi cluster network policies.
+The CRD will be namespace scoped (same as the Kubernetes NetworkPolicy resource).
+The CRD will have the same structure and fields as the default Kubernetes [NetworkPolicy],
+with additional fields that are relevant to the clusterset.
+
+The policies are local to the cluster, meaning they affect only the pods they are targeting on that cluster.
+If users want, they can create the same policy across different clusters manually.
+In the future, we may consider making policies global (or adding such a capability via a new CRD),
+allowing the users to create the policy once, and the system will take care of distributing it.
+
+Each such policy will be translated by the local CoastGuard controller to one (or more) `NetworkPolicy` where:
+
+* Local peers use the selector defined in the `ClusterSetNetworkPolicy` CR.
+* Remote peers (pods) are identified using an `ipBlock` with a `/32` CIDR.
```

Comment:
I hope that the initial implementation only supports IPv4 addresses. Please add a corresponding note.